File: src/main/java/com/mpe85/grampa/matcher/impl/TestNotRule.java
package com.mpe85.grampa.matcher.impl;

import com.mpe85.grampa.matcher.Rule;
import com.mpe85.grampa.matcher.RuleContext;

public class TestNotRule<T> extends AbstractRule<T> {
	
	public TestNotRule(final Rule<T> matcher) {
		super(matcher);
	}
	
	@Override
	public boolean match(final RuleContext<T> context) {
		final int currentIndex = context.getCurrentIndex();
		context.getValueStack().takeSnapshot();
		
		if (!context.getChildContext(getChild()).run()) {
			context.setCurrentIndex(currentIndex);
			context.getValueStack().restoreSnapshot();
			return true;
		}
		return false;
	}
	
	@Override
	public boolean isPredicate() {
		return true;
	}
	
}

File: src/main/kotlin/com/github/mpe85/grampa/rule/TestNotRule.kt
package com.github.mpe85.grampa.rule

import com.github.mpe85.grampa.context.ParserContext
import com.github.mpe85.grampa.util.checkEquality
import com.github.mpe85.grampa.util.stringify
import java.util.Objects.hash

/**
 * A predicate rule implementation that tests if its child rule does not match.
 *
 * @param[T] The type of the stack elements
 * @param[rule] The child rule to test
 * @author mpe85
 */
internal class TestNotRule<T>(rule: Rule<T>) : AbstractRule<T>(rule) {

    private val rule
        get() = checkNotNull(child)

    override val testRule: Boolean = true

    override fun match(context: ParserContext<T>): Boolean = !context.createChildContext(rule).run()

    override fun hashCode(): Int = hash(super.hashCode())

    override fun equals(other: Any?): Boolean = checkEquality(other, { super.equals(other) })

    override fun toString(): String = stringify("rule" to rule::class.simpleName)
}
